# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [1]:
# TODO: import the pandas library
import pandas as pd
# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', dtype=str)
df_electricity = pd.read_csv('electricity_access_percent.csv')

ParserError: Error tokenizing data. C error: Expected 3 fields in line 5, saw 63


In [3]:
# Investigate datasets
def print_lines(n, file):
    f = open(file)
    for line in range(n):
        print(f.readline())
    f.close()

In [5]:
print_lines(6, 'rural_population_percent.csv')

"Data Source","World Development Indicators",



"Last Updated Date","2018-06-28",



"Country Name","Country Code","Indicator Name","Indicator Code","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017",

"Aruba","ABW","Rural population (% of total population)","SP.RUR.TOTL.ZS","49.224","49.239","49.254","49.27","49.285","49.3","49.315","49.33","49.346","49.361","49.376","49.391","49.407","49.422","49.437","49.452","49.468","49.483","49.498","49.513","49.528","49.544","49.559","49.574","49.589","49.605","49.62","49.635","49.65","49.665","49.681","49.696","50.002","50.412","50.823","51.233","51.644","52.054","52.464","52.873","53.283","53.

In [6]:
print_lines(6, 'electricity_access_percent.csv')

"Data Source","World Development Indicators",



"Last Updated Date","2018-06-28",



"Country Name","Country Code","Indicator Name","Indicator Code","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017",

"Aruba","ABW","Access to electricity (% of population)","EG.ELC.ACCS.ZS","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","88.4453506469727","88.7808456420898","89.1158294677734","89.44775390625","89.7735595703125","90.0901870727539","90.3945846557617","90.6836776733398","90.9544219970703","91.2037506103516","91.6603980031457","91.6380920410156","91.8337173461914","92.0230484008789","92.2121658325

In [7]:
# Skip first 4 rows based on file investigation
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)

In [12]:
df_rural.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', 'Unnamed: 62'],
      dtype='object')

In [20]:
# TODO: remove the 'Unnamed:62' column from each data set
df_rural = df_rural.drop('Unnamed: 62', axis=1)
df_electricity = df_electricity.drop('Unnamed: 62', axis=1)

In [23]:
print(df_rural.shape)
print(df_rural.shape)

(264, 62)
(264, 62)


In [21]:
# TODO: combine the two data sets together using the concat method
df_rural_electricity = pd.concat([df_rural, df_electricity])
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.

In [22]:
df_rural_electricity

Country Name Country Code  \
0                                                Aruba          ABW   
1                                          Afghanistan          AFG   
2                                               Angola          AGO   
3                                              Albania          ALB   
4                                              Andorra          AND   
5                                           Arab World          ARB   
6                                 United Arab Emirates          ARE   
7                                            Argentina          ARG   
8                                              Armenia          ARM   
9                                       American Samoa          ASM   
10                                 Antigua and Barbuda          ATG   
11                                           Australia          AUS   
12                                             Austria          AUT   
13                                          Azerbaijan          AZE   
14                                             Burundi          BDI   
15                                             Belgium          BEL   
16                                               Benin          BEN   
17                                        Burkina Faso          BFA   
18                                          Bangladesh          BGD   
19                                            Bulgaria          BGR   
20                                             Bahrain          BHR   
21                                        Bahamas, The          BHS   
22                              Bosnia and Herzegovina          BIH   
23                                             Belarus          BLR   
24                                              Belize          BLZ   
25                                             Bermuda          BMU   
26                                             Bolivia          BOL   
27                                              Brazil          BRA   
28                                            Barbados          BRB   
29                                   Brunei Darussalam          BRN   
..                                                 ...          ...   
234  Latin America & the Caribbean (IDA & IBRD coun...          TLA   
235                                        Timor-Leste          TLS   
236  Middle East & North Africa (IDA & IBRD countries)          TMN   
237                                              Tonga          TON   
238                            South Asia (IDA & IBRD)          TSA   
239          Sub-Saharan Africa (IDA & IBRD countries)          TSS   
240                                Trinidad and Tobago          TTO   
241                                            Tunisia          TUN   
242                                             Turkey          TUR   
243                                             Tuvalu          TUV   
244                                           Tanzania          TZA   
245                                             Uganda          UGA   
246                                            Ukraine          UKR   
247                                Upper middle income          UMC   
248                                            Uruguay          URY   
249                                      United States          USA   
250                                         Uzbekistan          UZB   
251                     St. Vincent and the Grenadines          VCT   
252                                      Venezuela, RB          VEN   
253                             British Virgin Islands          VGB   
254                              Virgin Islands (U.S.)          VIR   
255                                            Vietnam          VNM   
256                                            Vanuatu          VUT   
257                                              World          WLD   
258                                              Samoa          WS

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [24]:
df_rural.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017'],
      dtype='object')

In [34]:
df_rural

Country Name Country Code  \
0                                                Aruba          ABW   
1                                          Afghanistan          AFG   
2                                               Angola          AGO   
3                                              Albania          ALB   
4                                              Andorra          AND   
5                                           Arab World          ARB   
6                                 United Arab Emirates          ARE   
7                                            Argentina          ARG   
8                                              Armenia          ARM   
9                                       American Samoa          ASM   
10                                 Antigua and Barbuda          ATG   
11                                           Australia          AUS   
12                                             Austria          AUT   
13                                          Azerbaijan          AZE   
14                                             Burundi          BDI   
15                                             Belgium          BEL   
16                                               Benin          BEN   
17                                        Burkina Faso          BFA   
18                                          Bangladesh          BGD   
19                                            Bulgaria          BGR   
20                                             Bahrain          BHR   
21                                        Bahamas, The          BHS   
22                              Bosnia and Herzegovina          BIH   
23                                             Belarus          BLR   
24                                              Belize          BLZ   
25                                             Bermuda          BMU   
26                                             Bolivia          BOL   
27                                              Brazil          BRA   
28                                            Barbados          BRB   
29                                   Brunei Darussalam          BRN   
..                                                 ...          ...   
234  Latin America & the Caribbean (IDA & IBRD coun...          TLA   
235                                        Timor-Leste          TLS   
236  Middle East & North Africa (IDA & IBRD countries)          TMN   
237                                              Tonga          TON   
238                            South Asia (IDA & IBRD)          TSA   
239          Sub-Saharan Africa (IDA & IBRD countries)          TSS   
240                                Trinidad and Tobago          TTO   
241                                            Tunisia          TUN   
242                                             Turkey          TUR   
243                                             Tuvalu          TUV   
244                                           Tanzania          TZA   
245                                             Uganda          UGA   
246                                            Ukraine          UKR   
247                                Upper middle income          UMC   
248                                            Uruguay          URY   
249                                      United States          USA   
250                                         Uzbekistan          UZB   
251                     St. Vincent and the Grenadines          VCT   
252                                      Venezuela, RB          VEN   
253                             British Virgin Islands          VGB   
254                              Virgin Islands (U.S.)          VIR   
255                                            Vietnam          VNM   
256                                            Vanuatu          VUT   
257                                              World          WLD   
258                                              Samoa          WS

In [41]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural_melt = pd.melt(df_rural, id_vars=['Country Name', 'Country Code'], 
       value_vars=list(df_rural.columns[4:]), 
       var_name='Year', value_name='Rural Population')

In [43]:
df_electricity_melt = pd.melt(df_electricity, id_vars=['Country Name', 'Country Code'], 
       value_vars=list(df_electricity.columns[4:]), 
       var_name='Year', value_name='Electricity Access')

In [53]:
# TODO: drop any columns from the data frames that aren't needed
# DONE in the melting phase

In [59]:
# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_combined = df_rural_melt.merge(df_electricity_melt, on=['Country Name', 'Country Code', 'Year'])

In [60]:
# TODO: sort the results by country and then by year
df_combined.sort_values(['Country Name', 'Year'], inplace=True)

In [61]:
df_combined

Country Name Country Code  Year  Rural Population  Electricity Access
1      Afghanistan          AFG  1960            91.779                 NaN
265    Afghanistan          AFG  1961            91.492                 NaN
529    Afghanistan          AFG  1962            91.195                 NaN
793    Afghanistan          AFG  1963            90.890                 NaN
1057   Afghanistan          AFG  1964            90.574                 NaN
1321   Afghanistan          AFG  1965            90.250                 NaN
1585   Afghanistan          AFG  1966            89.915                 NaN
1849   Afghanistan          AFG  1967            89.570                 NaN
2113   Afghanistan          AFG  1968            89.214                 NaN
2377   Afghanistan          AFG  1969            88.848                 NaN
2641   Afghanistan          AFG  1970            88.471                 NaN
2905   Afghanistan          AFG  1971            88.083                 NaN
3169   Afghanistan          AFG  1972            87.684                 NaN
3433   Afghanistan          AFG  1973            87.274                 NaN
3697   Afghanistan          AFG  1974            86.851                 NaN
3961   Afghanistan          AFG  1975            86.417                 NaN
4225   Afghanistan          AFG  1976            85.971                 NaN
4489   Afghanistan          AFG  1977            85.513                 NaN
4753   Afghanistan          AFG  1978            85.042                 NaN
5017   Afghanistan          AFG  1979            84.565                 NaN
5281   Afghanistan          AFG  1980            84.319                 NaN
5545   Afghanistan          AFG  1981            84.070                 NaN
5809   Afghanistan          AFG  1982            83.818                 NaN
6073   Afghanistan          AFG  1983            83.563                 NaN
6337   Afghanistan          AFG  1984            83.304                 NaN
6601   Afghanistan          AFG  1985            83.042                 NaN
6865   Afghanistan          AFG  1986            82.777                 NaN
7129   Afghanistan          AFG  1987            82.509                 NaN
7393   Afghanistan          AFG  1988            82.237                 NaN
7657   Afghanistan          AFG  1989            81.962                 NaN
...            ...          ...   ...               ...                 ...
7655      Zimbabwe          ZWE  1988            72.476                 NaN
7919      Zimbabwe          ZWE  1989            71.750                 NaN
8183      Zimbabwe          ZWE  1990            71.012           30.357832
8447      Zimbabwe          ZWE  1991            70.262           30.718805
8711      Zimbabwe          ZWE  1992            69.501           28.200000
8975      Zimbabwe          ZWE  1993            69.060           31.436678
9239      Zimbabwe          ZWE  1994            68.665           28.100000
9503      Zimbabwe          ZWE  1995            68.268           32.130077
9767      Zimbabwe          ZWE  1996            67.868           32.459949
10031     Zimbabwe          ZWE  1997            67.466           32.774529
10295     Zimbabwe          ZWE  1998            67.061           33.070751
10559     Zimbabwe          ZWE  1999            66.653           38.400000
10823     Zimbabwe          ZWE  2000            66.242           33.597420
11087     Zimbabwe          ZWE  2001            65.830           33.830868
11351     Zimbabwe          ZWE  2002            65.415           34.200000
11615     Zimbabwe          ZWE  2003            65.521           34.266785
11879     Zimbabwe          ZWE  2004            65.706           34.481380
12143     Zimbabwe          ZWE  2005            65.890           34.701820
12407     Zimbabwe          ZWE  2006            66.074           37.200000
12671     Zimbabwe          ZWE  2007            66.257           35.184479
12935     Zimbabwe          ZWE  2008            66.44